File: ZezexJava/server/src/Zezex/ProviderModuleBinds.java
package Zezex;

import Zeze.Services.ServiceManager.SubscribeInfo;
import Zezex.Provider.BModule;
import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;

import javax.xml.parsers.DocumentBuilderFactory;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.*;
import java.io.*;

public class ProviderModuleBinds {

	public static ProviderModuleBinds Load() {
		return Load(null);
	}

//C# TO JAVA CONVERTER NOTE: Java does not support optional parameters. Overloaded method(s) are created above:
//ORIGINAL LINE: public static ProviderModuleBinds Load(string xmlfile = null)
	public static ProviderModuleBinds Load(String xmlfile) {
		if (xmlfile == null) {
			xmlfile = "provider.module.binds.xml";
		}

		if (Files.isRegularFile(Paths.get(xmlfile))) {
			DocumentBuilderFactory db = DocumentBuilderFactory.newInstance();
			db.setXIncludeAware(true);
			db.setNamespaceAware(true);
			try {
				Document doc = db.newDocumentBuilder().parse(xmlfile);
				return new ProviderModuleBinds(doc.getDocumentElement());
			}
			catch (Exception ex) {
				throw new RuntimeException(ex);
			}
		}
		return new ProviderModuleBinds();
	}

	private boolean IsDynamicModule(String name) {
		var m = Modules.get(name);
		if (null != m) {
			return m.Providers.isEmpty();
		}
		return false;
	}

	private int GetModuleChoiceType(String name) {
		var m = Modules.get(name);
		if (null != m) {
			return m.ChoiceType;
		}
		return Zezex.Provider.BModule.ChoiceTypeDefault;
	}

	public final void BuildDynamicBinds(HashMap<String, Zeze.IModule> AllModules, int serverId, HashMap<Integer, Zezex.Provider.BModule> out) {
		for (var m : AllModules.values()) {
			var cm = Modules.get(m.getFullName());
			if (null == cm)
				continue; // dynamic must exist in config
			if (cm.getConfigType() != BModule.ConfigTypeDynamic)
				continue;

			if (!cm.Providers.isEmpty() && !cm.Providers.contains(serverId))
				continue; // dynamic providers. isEmpty means enable in all server.

			var tempVar = new Zezex.Provider.BModule();
			tempVar.setChoiceType(cm.getChoiceType());
			tempVar.setConfigType(BModule.ConfigTypeDynamic);
			tempVar.setSubscribeType(cm.getSubscribeType());
			out.put(m.getId(), tempVar);
		}
	}

	public final void BuildStaticBinds(HashMap<String, Zeze.IModule> AllModules, int serverId, HashMap<Integer, Zezex.Provider.BModule> modules) {
		HashMap<String, Integer> binds = new HashMap<String, Integer>();

		// special binds
		for (var m : getModules().values()) {
			if (m.getConfigType() == BModule.ConfigTypeSpecial && m.Providers.contains(serverId)) {
				binds.put(m.FullName, Zezex.Provider.BModule.ConfigTypeSpecial);
			}
		}

		// default binds
		if (false == getProviderNoDefaultModule().contains(serverId)) {
			for (var m : AllModules.values()) {
				if (IsDynamicModule(m.getFullName())) {
					continue; // 忽略动态注册的模块。
				}
				if (getModules().containsKey(m.getFullName())) {
					continue; // 忽略已经有特别配置的模块
				}
				binds.put(m.getFullName(), Zezex.Provider.BModule.ConfigTypeDefault);
			}
		}

		// output
		for (var bind : binds.entrySet()) {
			var m = AllModules.get(bind.getKey());
			if (null != m){
				Zezex.Provider.BModule tempVar = new Zezex.Provider.BModule();
				tempVar.setChoiceType(GetModuleChoiceType(bind.getKey()));
				tempVar.setConfigType(bind.getValue());
				tempVar.setSubscribeType(Zeze.Services.ServiceManager.SubscribeInfo.SubscribeTypeReadyCommit);
				modules.put(m.getId(), tempVar);
			}
		}
	}

	public static class Module {
		private String FullName;
		public final String getFullName() {
			return FullName;
		}
		private int ChoiceType;
		public final int getChoiceType() {
			return ChoiceType;
		}
		private int SubscribeType;
		public final int getSubscribeType() { return SubscribeType; }
		private int ConfigType; // 为了兼容，没有配置的话，从其他条件推导出来。
		public final int getConfigType() { return ConfigType; }

		private HashSet<Integer> Providers = new HashSet<Integer> ();
		public final HashSet<Integer> getProviders() {
			return Providers;
		}

		private int GetChoiceType(Element self) {
			switch (self.getAttribute("ChoiceType")) {
				case "ChoiceTypeHashAccount":
					return Zezex.Provider.BModule.ChoiceTypeHashAccount;

				case "ChoiceTypeHashRoleId":
					return Zezex.Provider.BModule.ChoiceTypeHashRoleId;

				default:
					return Zezex.Provider.BModule.ChoiceTypeDefault;
			}
		}

		// 这个订阅类型目前用于动态绑定的模块，所以默认为SubscribeTypeSimple。
		private int GetSubscribeType(Element self) {
			switch (self.getAttribute("SubscribeType")) {
				case "SubscribeTypeReadyCommit":
					return SubscribeInfo.SubscribeTypeReadyCommit;
				//case "SubscribeTypeSimple":
				//	return SubscribeInfo.SubscribeTypeSimple;
				default:
					return SubscribeInfo.SubscribeTypeSimple;
			}
		}

		public Module(Element self) {
			FullName = self.getAttribute("name");
			ChoiceType = GetChoiceType(self);
			SubscribeType = GetSubscribeType(self);

			ProviderModuleBinds.SplitIntoSet(self.getAttribute("providers"), getProviders());

			String attr = self.getAttribute("ConfigType").trim();
			switch (attr) {
				case "":
					// 兼容，如果没有配置
					ConfigType = Providers.isEmpty() ? BModule.ConfigTypeDynamic : BModule.ConfigTypeSpecial;
					break;

				case "Special":
					ConfigType = BModule.ConfigTypeSpecial;
					break;

				case "Dynamic":
					ConfigType = BModule.ConfigTypeDynamic;
					break;

				case "Default":
					ConfigType = BModule.ConfigTypeDefault;
					break;

				default:
					throw new RuntimeException("unknown ConfigType " + attr);
			}
		}
	}

	private HashMap<String, Module> Modules = new HashMap<String, Module> ();
	public final HashMap<String, Module> getModules() {
		return Modules;
	}
	private HashSet<Integer> ProviderNoDefaultModule = new HashSet<Integer> ();
	public final HashSet<Integer> getProviderNoDefaultModule() {
		return ProviderNoDefaultModule;
	}

	private ProviderModuleBinds() {
	}

	private ProviderModuleBinds(Element self) {
		if (false == self.getNodeName().equals("ProviderModuleBinds")) {
			throw new RuntimeException("is it a ProviderModuleBinds config.");
		}

		NodeList childnodes = self.getChildNodes();
		for (int i = 0; i < childnodes.getLength(); ++i) {
			Node node = childnodes.item(i);
			if (Node.ELEMENT_NODE != node.getNodeType()) {
				continue;
			}

			Element e = (Element)node;
			switch (e.getNodeName()) {
				case "module":
					AddModule(new Module(e));
					break;

				case "ProviderNoDefaultModule":
					SplitIntoSet(e.getAttribute("providers"), getProviderNoDefaultModule());
					break;

				default:
					throw new RuntimeException("unknown node name: " + e.getNodeName());
			}
		}
	}

	private void AddModule(Module module) {
		getModules().put(module.getFullName(), module);
	}

	private static void SplitIntoSet(String providers, HashSet<Integer> set) {
		for (var provider : providers.split("[,]", -1)) {
			var p = provider.trim();
			if (p.isEmpty())
				continue;
			set.add(Integer.parseInt(p));
		}
	}
}

File: Zeze/Arch/ProviderModuleBinds.cs
using System;
using System.Collections.Generic;
using System.Xml;
using Zeze.Builtin.Provider;

namespace Zeze.Arch
{
    public class ProviderModuleBinds
    {
        public static ProviderModuleBinds Load(string xmlfile = null)
        {
            if (xmlfile == null)
                xmlfile = "provider.module.binds.xml";

            if (System.IO.File.Exists(xmlfile))
            {
                XmlDocument doc = new XmlDocument();
                doc.Load(xmlfile);
                return new ProviderModuleBinds(doc.DocumentElement);
            }
            return new ProviderModuleBinds();
        }

        private bool IsDynamicModule(string name)
        {
            if (Modules.TryGetValue(name,out var m))
            {
                return m.Providers.Count == 0;
            }
            return false;
        }

        private int GetModuleChoiceType(string name)
        {
            if (Modules.TryGetValue(name, out var m))
            {
                return m.ChoiceType;
            }
            return BModule.ChoiceTypeDefault;
        }

        public void BuildDynamicBinds(Dictionary<string, IModule> AllModules, int serverId, Dictionary<int, BModule> result)
        {
            foreach (var m in AllModules)
            {
                if (Modules.TryGetValue(m.Value.FullName, out var cm))
                {
                    if (!cm.Dynamic)
                        continue;

                    if (cm.Providers.Count > 0 && !cm.Providers.Contains(serverId))
                        continue; // dynamic providers. isEmpty means enable in all server.

                    var tempVar = new BModule();
                    tempVar.ChoiceType = cm.ChoiceType;
                    tempVar.Dynamic = true;
                    result.Add(m.Value.Id, tempVar);
                }
            }
        }

        public void BuildStaticBinds(Dictionary<string, IModule> AllModules,
            int serverId, Dictionary<int, BModule> modules)
        {
            Dictionary<string, bool> binds = new Dictionary<string, bool>();

            // special binds
            foreach (var m in Modules.Values)
            {
                if (!m.Dynamic && m.Providers.Contains(serverId))
                    binds.Add(m.FullName, false);
            }

            // default binds
            if (false == ProviderNoDefaultModule.Contains(serverId))
            {
                foreach (var m in AllModules.Values)
                {
                    if (IsDynamicModule(m.FullName))
                        continue; // 忽略动态注册的模块。
                    if (Modules.ContainsKey(m.FullName))
                        continue; // 忽略已经有特别配置的模块
                    binds.Add(m.FullName, false);
                }
            }

            // output
            foreach (var bind in binds)
            {
                if (AllModules.TryGetValue(bind.Key, out var m))
                    modules.Add(m.Id, new BModule()
                    {
                        ChoiceType = GetModuleChoiceType(bind.Key),
                        Dynamic = bind.Value,
                    });
            }
        }

        public class Module
        { 
            public string FullName { get; }
            public int ChoiceType { get; }
            public HashSet<int> Providers { get; } = new HashSet<int>();

            public bool Dynamic { get; }

            private int GetChoiceType(XmlElement self)
            {
                switch (self.GetAttribute("ChoiceType"))
                {
                    case "ChoiceTypeHashAccount":
                        return BModule.ChoiceTypeHashAccount;

                    case "ChoiceTypeHashRoleId":
                        return BModule.ChoiceTypeHashRoleId;

                    default:
                        return BModule.ChoiceTypeDefault;
                }
            }

            public Module(XmlElement self)
            {
                FullName = self.GetAttribute("name");
                ChoiceType = GetChoiceType(self);

                ProviderModuleBinds.SplitIntoSet(self.GetAttribute("providers"), Providers);

                string attr = self.GetAttribute("dynamic").Trim();
                Dynamic = !string.IsNullOrEmpty(attr) && bool.Parse(attr);
            }
        }

        public Dictionary<string, Module> Modules { get; } = new Dictionary<string, Module>();
        public HashSet<int> ProviderNoDefaultModule { get; } = new HashSet<int>();

        private ProviderModuleBinds()
        { 
        }

        private ProviderModuleBinds(XmlElement self)
        {
            if (false == self.Name.Equals("ProviderModuleBinds"))
                throw new Exception("is it a ProviderModuleBinds config.");

            XmlNodeList childNodes = self.ChildNodes;
            foreach (XmlNode node in childNodes)
            {
                if (XmlNodeType.Element != node.NodeType)
                    continue;

                XmlElement e = (XmlElement)node;
                switch (e.Name)
                {
                    case "module":
                        AddModule(new Module(e));
                        break;

                    case "ProviderNoDefaultModule":
                        SplitIntoSet(e.GetAttribute("providers"), ProviderNoDefaultModule);
                        break;

                    default:
                        throw new Exception("unknown node name: " + e.Name);
                }
            }
        }

        private void AddModule(Module module)
        {
            Modules.Add(module.FullName, module);
        }

        private static void SplitIntoSet(string providers, HashSet<int> set)
        {
            foreach (var provider in providers.Split(','))
            {
                var p = provider.Trim();
                if (p.Length == 0)
                    continue;
                set.Add(int.Parse(p));
            }
        }
    }
}
